components/script/dom/htmloptionelement.rs

```diff
@@ -108,51 +103,66 @@ impl HTMLOptionElement {
 
     // https://html.spec.whatwg.org/multipage/#concept-option-index
     fn index(&self) -> i32 {
-        if let Some(parent) = self.upcast::<Node>().GetParentNode() {
-            if let Some(select_parent) = parent.downcast::<HTMLSelectElement>() {
-                // return index in parent select's list of options
-                return self.index_in_select(select_parent);
-            } else if parent.is::<HTMLOptGroupElement>() {
-                if let Some(grandparent) = parent.GetParentNode() {
-                    if let Some(select_grandparent) = grandparent.downcast::<HTMLSelectElement>() {
-                        // return index in grandparent select's list of options
-                        return self.index_in_select(select_grandparent);
-                    }
-                }
+        if let Some(owner_select) = self.owner_select_element() {
+            match owner_select.list_of_options().position(|n| &*n == self) {
+                Some(index) => return index.try_into().unwrap_or(0),
+                None => {
+                    // shouldn't happen but not worth a browser panic
+                    warn!(
+                        "HTMLOptionElement called index_in_select at a select that did not contain it"
+                    );
+                    return 0;
+                },
             }
         }
+
         // "If the option element is not in a list of options,
         // then the option element's index is zero."
         // self is neither a child of a select, nor a grandchild of a select
         // via an optgroup, so it is not in a list of options
         0
     }
 
-    fn index_in_select(&self, select: &HTMLSelectElement) -> i32 {
-        match select.list_of_options().position(|n| &*n == self) {
-            Some(index) => index.try_into().unwrap_or(0),
-            None => {
-                // shouldn't happen but not worth a browser panic
-                warn!(
-                    "HTMLOptionElement called index_in_select at a select that did not contain it"
-                );
-                0
-            },
+    fn owner_select_element(&self) -> Option<DomRoot<HTMLSelectElement>> {
+        if let Some(parent) = self.upcast::<Node>().GetParentNode() {
+            if parent.is::<HTMLOptGroupElement>() {
+                if let Some(grandparent) = parent.GetParentNode() {
+                    DomRoot::downcast::<HTMLSelectElement>(grandparent)
+                } else {
+                    None
+                }
+            } else {
+                DomRoot::downcast::<HTMLSelectElement>(parent)
+            }
+        } else {
+            None
         }
```

Comment:
Could be a little simpler like this:

```suggestion
        let parent = self.upcast::<Node>().GetParentNode()?;
        if  parent.is::<HTMLOptGroupElement>() {
            return None;
        }
       DomRoot::downcast::<HTMLSelectElement>(parent.GetParentNode()?)
```

Reply:
You're right, I should be using the `?` operator, but I don't think your suggestion is correct.

`if  parent.is::<HTMLOptGroupElement>() {` should probably be negated and the owning select might also be the direct parent of the `<option>`.

Reply:
Changed it to 
```rust
let parent = self.upcast::<Node>().GetParentNode()?;

if parent.is::<HTMLOptGroupElement>() {
     DomRoot::downcast::<HTMLSelectElement>(parent.GetParentNode()?)
} else {
     DomRoot::downcast::<HTMLSelectElement>(parent)
}
```